src/tracing.c

```diff
@@ -0,0 +1,204 @@
+/****************************************************************************
+**
+**  This file is part of GAP, a system for computational discrete algebra.
+**
+**  Copyright of GAP belongs to its developers, whose names are too numerous
+**  to list here. Please refer to the COPYRIGHT file for details.
+**
+**  SPDX-License-Identifier: GPL-2.0-or-later
+**
+**  This file contains functionality for tracing operations
+**
+*/
+
+#include "tracing.h"
+
+#include "bool.h"
+#include "integer.h"
+#include "lists.h"
+#include "modules.h"
+#include "plist.h"
+#include "precord.h"
+#include "records.h"
+
+// This file contains functionality which stores how often internal operations
+// (like +, ^, -) are executed on each TNUM.
+
+
+static Obj RecordedStats;
+
+// Report the 1 argument operation 'name' was applied to 'op'
+void ReportWrappedOperation1(const char * cname, Obj op)
+{
+    UInt name = RNamName(cname);
+    if (!ISB_REC(RecordedStats, name)) {
+        Obj list = NEW_PLIST(T_PLIST, 0);
+        ASS_REC(RecordedStats, name, list);
+    }
+    Obj list = ELM_REC(RecordedStats, name);
+
+    UInt tnam = TNUM_OBJ(op) + 1;
+    Obj  val = ELM0_LIST(list, tnam);
+    if (!val) {
+        val = INTOBJ_INT(0);
+    }
+
+    UInt8 intval = Int8_ObjInt(val);
+    intval++;
+    val = ObjInt_Int8(intval);
+
+    ASS_LIST(list, tnam, val);
+}
+
+// Report the 2 argument operation 'name' was applied to 'op1' and 'op2'
+void ReportWrappedOperation2(const char * cname, Obj op1, Obj op2)
+{
+    UInt name = RNamName(cname);
+    if (!ISB_REC(RecordedStats, name)) {
+        Obj list = NEW_PLIST(T_PLIST, 0);
+        ASS_REC(RecordedStats, name, list);
+    }
+    Obj list = ELM_REC(RecordedStats, name);
+
+    UInt tnam1 = TNUM_OBJ(op1) + 1;
+    Obj  pos = ELM0_LIST(list, tnam1);
+    if (!pos) {
+        pos = NEW_PLIST(T_PLIST, 0);
+        ASS_LIST(list, tnam1, pos);
+    }
+
+    UInt tnam2 = TNUM_OBJ(op2) + 1;
+    Obj  val = ELM0_LIST(pos, tnam2);
+    if (!val) {
+        val = INTOBJ_INT(0);
+    }
+
+    UInt8 intval = Int8_ObjInt(val);
+    intval++;
+    val = ObjInt_Int8(intval);
+
+    ASS_LIST(pos, tnam2, val);
+}
+
+typedef struct {
+    voidfunc activate;
+    voidfunc deactivate;
+} voidfuncs;
+
+// Store the list of operators which can have tracing enabled and disabled
+static voidfuncs controllers[64];
+static int       tracking_active;
+
+void InstallOpWrapper(voidfunc activate, voidfunc deactivate)
+{
+    int pos = 0;
+    while (pos < 64 && controllers[pos].activate != 0) {
```

Comment:
Instead of hardcoding 64 here and elsewhere, it would be better to use `ARRAYSIZE(controllers)`. Or use a global enum for the size. Or don't use a size at all, and instead use a PLIST to store this data.

Reply:
I started switching to PLIST, but it doesn't really work as we have to store C function pointers